### PR TITLE
Add --sandbox-canton option in daml start

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Part of #11831

This is an intermediate step to dropping the other versions of sandbox.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs [hunk 1]
```diff
@@ -168,9 +168,22 @@ commandParser = subparser $ fold
         jsonApiOptions <- many (strOption (long "json-api-option" <> metavar "JSON_API_OPTION" <> help "Pass option to HTTP JSON API"))
         scriptOptions <- many (strOption (long "script-option" <> metavar "SCRIPT_OPTION" <> help "Pass option to Daml script interpreter"))
         shutdownStdinClose <- stdinCloseOpt
-        sandboxClassic <- SandboxClassic <$> switch (long "sandbox-classic" <> help "Deprecated. Run with Sandbox Classic.")
+        sandboxChoice <- sandboxChoiceOpt
+        cantonAdminApiPort <- sandboxPortOpt "canton-admin-api-port" "Port number for the canton admin API (--sandbox-canton only)"
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:172]: should we setup optparse-applicative to allow them only in combination? afaict in the CLI parser you do accept them even if you don’t use the canton sandbox.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs [hunk 2]
```diff
@@ -168,9 +168,22 @@ commandParser = subparser $ fold
...
 
+    sandboxPortOpt name desc =
+        optional (option (maybeReader (toSandboxPortSpec <=< readMaybe))
+            (long name <> metavar "PORT_NUM" <> help desc))
+
+    sandboxChoiceOpt =
+            SandboxClassic <$ switch (long "sandbox-classic" <> help "Deprecated. Run with Sandbox Classic.")
+        <|> SandboxModern <$ switch (long "sandbox-modern" <> help "Deprecated. Run with pre-2.0 Sandbox.")
+        <|> SandboxCanton <$ switch (long "sandbox-canton" <> help "Run with Canton Sandbox. The 2.0 default.")
+        <|> pure SandboxModern -- pre-2.0 default
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:185]: I’d suggest renaming `SandboxModern` to `SandboxKV`. A bit more explicit.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 3]
```diff
@@ -95,20 +95,53 @@ navigatorPortNavigatorArgs (NavigatorPort p) = ["--port", show p]
...
+      SandboxModern -> oldSandbox "sandbox"
+      SandboxCanton -> cantonSandbox
+
+  where
+    cantonSandbox = do
+      cantonPorts <- determineCantonPortsLabouriously startOptions
+      withCantonSandbox cantonPorts sandboxArgs $ \ph -> do
+        let sandboxPort = ledgerApi cantonPorts
+        putStrLn "Waiting for canton sandbox to start: "
+        waitForConnectionOnPort (putStr "." *> threadDelay 500000) sandboxPort
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:127]: Not part of this PR but this is unreliable afaict because it can happen before the participant has connected to the domain. I suggest asking the Canton team what they recommend here.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Util.hs [hunk 4]
```diff
@@ -288,3 +288,26 @@ cantonConfig CantonPorts{..} =
   where
     port p = Aeson.object [ "port" Aeson..= p ]
     storage = "storage" Aeson..= Aeson.object [ "type" Aeson..= ("memory" :: T.Text) ]
+
+-- This is slightly hacky: we need to find a free port but pass it to an
```
Comment (akrmn) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Util.hs:292]: I think it's just the right amount of hacky :) thanks for the comment!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 5]
```diff
@@ -95,20 +95,53 @@ navigatorPortNavigatorArgs (NavigatorPort p) = ["--port", show p]
...
+        runLedgerUploadDar ((defaultLedgerFlags Grpc) {fPortM = Just sandboxPort}) (Just darPath)
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:128]: To make this explicit: I don’t think we need to support DARs on the CLI for `daml canton-sandbox` in 2.0 and can instead always go this way.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 6]
```diff
@@ -95,20 +95,53 @@ navigatorPortNavigatorArgs (NavigatorPort p) = ["--port", show p]
...
-          , concat [ [ "--port", show (fromSandboxPortSpec portSpec) ] | Just portSpec <- [mbPortSpec] ]
-          , [ "--port-file", portFile ]
-          , extraArgs
-          ]
-    withPlatformJar args "sandbox-logback.xml" $ \ph -> do
-        putStrLn "Waiting for sandbox to start: "
-        port <- readPortFile maxRetries portFile
-        a ph (SandboxPort port)
+-- | Use SandboxPortSpec to determine a sandbox port number.
+-- This is racy thanks to getFreePort, but there's no good alternative at the moment.
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:99]: Same as below, let’s reach out to the Canton team to ask for ideas/port files

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Util.hs [hunk 7]
```diff
@@ -288,3 +288,26 @@ cantonConfig CantonPorts{..} =
...
+-- external process. Technically this port could be reused between us
+-- getting it from the kernel and the external process listening
+-- on that port but ports are usually not reused aggressively so this should
+-- be fine and is certainly better than hardcoding the port.
+getFreePort :: IO PortNumber
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Util.hs:297]: arguably da-hs-base is a better place for this but both is fine with me.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 2 above)
Comment (cocreature) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:185]: maybe also good to add a TODO to switch the default here.

**Review by cocreature — APPROVED** [review]
Looks great, thank you! Maybe worth adding a test to the assistant integration tests?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 1 above)
Comment (akrmn) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs:172]: `mappend (Sum 1)`

**Review by akrmn — APPROVED** [review]
LGTM, thanks!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs (hunk 5 above)
Comment (sofiafaro-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:128]: It makes sense to have it here though, right?

Comment (cocreature) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:128]: Yeah absolutely, was just trying to suggest that we shouldn’t try to implement some magic for `daml canton-sandbox` that turns CLI args into a ledger upload.

Comment (sofiafaro-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:128]: 👍 that makes sense!

**Review by robin-da — APPROVED** [review]
Looks great, looking forward for this to land. Nothing more to add to what's already been commented.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs (hunk 1 above)
Comment (sofiafaro-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Main.hs]: Changed this to only have the canton port options with --sandbox-canton. This is cleaner.

I also realize I had a bug in the option parser (I was using `switch` instead of `flag'`, which is what I need to the alternative work like intended). This was causing sandbox-classic to be the default.

**Comment by sofiafaro-da** [comment]
> Maybe worth adding a test to the assistant integration tests?

Definitely, I'll work on this but might merge it as a separate PR.

File: ledger/ledger-api-test-tool-on-canton/canton.conf [hunk 8]
```diff
@@ -2,14 +2,6 @@
 # SPDX-License-Identifier: Apache-2.0
 
 canton {
-  parameters {
-    participant {
-      party-change-notification {
-        type = via-domain
-      }
-    }
-  }
```
Comment (sofiafaro-da) [inline ledger/ledger-api-test-tool-on-canton/canton.conf:11]: These canton.parameters.participant are now specified per participant, but "via-domain" is the default party-change-notification type anyway, so it seemed cleaner to just remove this.
